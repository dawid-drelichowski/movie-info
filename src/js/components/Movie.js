import React, {Component} from 'react'
import PropTypes from 'prop-types'
import MovieDetails from 'components/MovieDetails'

export default class Movie extends Component {
  static propTypes = {
    config: PropTypes.object.isRequired,
    service: PropTypes.object.isRequired,
    model: PropTypes.object.isRequired
  }
  state = {detailsVisible: false, details: null}
  getImageUrl () {
    const config = this.props.config.movieDb.images
    return `${config.url}${config.size.small}`
  }
  toggleDetails = async () => {
    const state = {detailsVisible: false}
    if (this.state.detailsVisible) {
      this.setState(state)
      return
    }
    state.detailsVisible = true
    if (!this.state.details) {
      state.details = await this.props.service.findById(this.props.model.id)
    }
    this.setState(state)
  }
  render () {
    const model = this.props.model
    return <li onClick={this.toggleDetails} className="list-group-item list-group-item-action movie">
      <strong>{model.title}</strong>
      {model.imagePath ? this.renderImage() : ''}
      {this.state.detailsVisible ? this.renderMovieDetails() : ''}
    </li>
  }
  renderImage () {
    const model = this.props.model
    return <img
      src={`${this.getImageUrl()}${model.imagePath}`}
      alt={model.title}
      className="rounded float-right"
    />
  }
  renderMovieDetails () {
    return <MovieDetails config={this.props.config} model={this.state.details}/>
  }
}
