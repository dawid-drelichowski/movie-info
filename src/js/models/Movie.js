export default class {
  id = 0
  title = ''
  imagePath = ''
}
