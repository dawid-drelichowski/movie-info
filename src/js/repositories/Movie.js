export default class {
  config = {}
  constructor (config) {
    this.config = config
  }
  async searchByQuery (query) {
    if (query) {
      return this.fetch(this.getSearchByQueryUrl(query))
    }
    return false
  }
  async findById (id) {
    if (id) {
      return this.fetch(this.getFindByIdUrl(id))
    }
    return false
  }
  async fetch (url) {
    try {
      const response = await fetch(url)
      if (response.ok) {
        return await response.json()
      }
      return false
    } catch (error) {
      console.log(error)
      return false
    }
  }
  getSearchByQueryUrl (query) {
    const config = this.getMovieDbConfig()
    return `${config.url}search/movie?api_key=${config.apiKey}&query=${query}`
  }
  getFindByIdUrl (id) {
    const config = this.getMovieDbConfig()
    return `${config.url}movie/${id}?api_key=${config.apiKey}`
  }
  getMovieDbConfig () {
    return this.config.movieDb
  }
}
