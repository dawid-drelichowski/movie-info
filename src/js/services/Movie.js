import Mapper from 'mappers/Movie'
import DetailsMapper from 'mappers/MovieDetails'

export default class {
  repository
  constructor (repository) {
    this.repository = repository
  }
  async searchByQuery (query) {
    const data = await this.repository.searchByQuery(query)
    if (data && data.results) {
      return Mapper.mapCollection(data.results)
    }
    return false
  }
  async findById (id) {
    const data = await this.repository.findById(id)
    if (data) {
      return DetailsMapper.map(data)
    }
    return false
  }
}
