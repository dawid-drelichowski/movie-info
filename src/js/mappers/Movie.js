import Abstract from 'mappers/Abstract'
import Movie from 'models/Movie'

export default class extends Abstract {
  static mapping = new Map([
    ['id', 'id'],
    ['title', 'title'],
    ['imagePath', 'backdrop_path']
  ])
  static map (data) {
    return this.performMapping(data, new Movie())
  }
}
