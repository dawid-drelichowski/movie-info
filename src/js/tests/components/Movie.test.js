import React from 'react'
import Movie from 'components/Movie'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme'
import config from 'tests/mocks/config.json'
import repository from 'tests/mocks/repository'
import search from 'tests/mocks/search.json'

describe('Movie component', () => {
  function getRenderedComponent (data) {
    return renderer.create(<Movie config={config} repository={repository} data={data}/>)
  }

  function getShallowComponent (data) {
    return shallow(<Movie config={config} repository={repository} data={data}/>)
  }

  it('should not have image', () => {
    expect(getRenderedComponent({...search.results[0], backdrop_path: undefined}).toJSON()).toMatchSnapshot()
  })

  it('should have image', () => {
    expect(getRenderedComponent(search.results[0]).toJSON()).toMatchSnapshot()
  })

  it('should have details', async () => {
    const wrapper = getShallowComponent(search.results[0])
    await wrapper.instance().toggleDetails()
    expect(wrapper.find('MovieDetails').exists()).toBeTruthy()
  })
})
