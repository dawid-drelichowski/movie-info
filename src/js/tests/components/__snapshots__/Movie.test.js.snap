// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie component should have image 1`] = `
<li
  className="list-group-item list-group-item-action movie"
  onClick={[Function]}
>
  <strong>
    Title
  </strong>
  <img
    alt="Title"
    className="rounded float-right"
    src="https://image.mock.api/small/image-path.jpg"
  />
  
</li>
`;

exports[`Movie component should not have image 1`] = `
<li
  className="list-group-item list-group-item-action movie"
  onClick={[Function]}
>
  <strong>
    Title
  </strong>
  
  
</li>
`;
